### PR TITLE
[rllib, train] Add support for nested metrics in `Result.get_best_checkpoint`

### PR DESCRIPTION
**Pull request by pseudo-rnd-thoughts**
## Description
RLlib uses nested metric structure (like `"{ENV_RUNNER_RESULTS}/{EPISODE_RETURN_MEAN}"`) which `Result.get_best_checkpoint` doesn't support.
Following `ResultGrid.get_best_result()` to use `unflattened_lookup`, I've added that to `get_best_checkpoint` along with testing for nested structures (and its backward compatibility) 

Reproduction script
```python
from ray import tune
from ray.rllib.algorithms.ppo import PPOConfig
from ray.rllib.core.rl_module.default_model_config import DefaultModelConfig
from ray.rllib.utils.metrics import (
    ENV_RUNNER_RESULTS,
    EPISODE_RETURN_MEAN,
    NUM_ENV_STEPS_SAMPLED_LIFETIME,
)
from ray.tune.result import TRAINING_ITERATION

config = (
    PPOConfig()
    .environment("CartPole-v1")
    .training(num_epochs=6)
)

tuner = tune.Tuner(
    "PPO",
    param_space=config.to_dict(),
    run_config=tune.RunConfig(
        "PPO_Reproduce",
        checkpoint_config=tune.CheckpointConfig(
            num_to_keep=10,
            checkpoint_score_attribute=f"{ENV_RUNNER_RESULTS}/{EPISODE_RETURN_MEAN}",
            checkpoint_at_end=True,
            checkpoint_frequency=5,
        ),
        stop={
            f"{ENV_RUNNER_RESULTS}/{NUM_ENV_STEPS_SAMPLED_LIFETIME}": 3e5,
            f"{ENV_RUNNER_RESULTS}/{EPISODE_RETURN_MEAN}": 450,
            TRAINING_ITERATION: 100,
        },
    ),
)

results = tuner.fit()
best_result = results.get_best_result()
ckpt = best_result.get_best_checkpoint(f"{ENV_RUNNER_RESULTS}/{EPISODE_RETURN_MEAN}", "max")
print(ckpt.path)
```

## Related issues
https://github.com/ray-project/ray/issues/57533

### CODE REVIEW

File: python/ray/air/result.py [hunk 1]
```diff
@@ -10,6 +10,7 @@
 import pyarrow
 
 import ray
+from ray._private.dict import unflattened_lookup
```
Comment (matthewdeng) [inline python/ray/air/result.py:13]: @justinvyu do we want to support this for Train V2 as well, or should we diverge for Tune?

Comment (Morphlng) [inline python/ray/air/result.py:13]: Is it a good idea to use a function that is marked `Deprecated`? Or should this mark be removed.

Comment (pseudo-rnd-thoughts) [reply python/ray/air/result.py:13]: Good point. Looking at the related PR there isn't a note on the future plan for the functions marked deprecated. Any suggestions on an alternative functions for the same job?

Comment (Morphlng) [reply python/ray/air/result.py:13]: Not sure about this. It seems like `ray.tune` has its dependencies to these utilities still, e.g. `ray.tune.progress_reporter.TuneReporterBase` and `ray.tune.analysis.experiment_analysis.ExperimentAnalysis`.

Comment (justinvyu) [reply python/ray/air/result.py:13]: I'm ok to support this in Train. This is only needed if users self-report nested dicts.

File: python/ray/train/v2/tests/test_result.py [hunk 2]
```diff
@@ -116,6 +116,90 @@ def test_get_best_checkpoint():
     )
 
 
+def test_get_best_checkpoint_nested_metrics():
+    """Test that get_best_checkpoint works with nested metric dictionaries.
+
+    RLlib uses nested metric structure like {"env_runners": {"episode_return_mean": value}}
```
Comment (justinvyu) [inline python/ray/train/v2/tests/test_result.py]:
```suggestion
```

**Review by justinvyu — APPROVED** [review]
Thanks!
